NotificationWorker for Realtime updates

gokmen opened the pull request:
## Description

This PR provides in house realtime notification support for all the existing events that we use with PubNub. 

## Motivation and Context

Currently we don't have realtime update support for `default` environment which is provided for on premise installs. To be able to have realtime support on on premise Koding you need to have a paid PubNub subscription but with this PR we're providing same feature set in house, by using existing stack that we have; rabbitmq, mongo and sockjs used for implementation.

Notification workers are designed to handle all the requirements to fill this gap and they are horizontally scalable for web socket usage. For long polling load balancer needs to be configure to have sticky session support. see https://github.com/koding/koding/commit/7bb93076049ca00773557eeb9f06d73f50124c85

Notification Workers does following;
 - [x] provides a REST api for drop-in-replacement with socialApi's existing dispatcher
 - [x] provides SockJS server for clients to connect for realtime updates
 - [x] consumes RabbitMQ for publishing and consuming realtime events
 - [x] on client side it has a simple sock-js client implementation with auto reconnect support with backoff

Can be shown as;

![2017-05-23 19-25-50](https://cloud.githubusercontent.com/assets/42368/26384622/77f50764-3fef-11e7-867a-522dbda93cff.png)

It provides user and team (group) notifications via different routes, to have same ability with existing channel implementation;

![2017-05-23 17-21-17](https://cloud.githubusercontent.com/assets/42368/26384671/b859c984-3fef-11e7-8ab7-b386dfdfce5d.png)

There is a simple status endpoint which shows the connection statuses that can be disabled from nginx configuration;

![2017-05-23 17-33-00](https://cloud.githubusercontent.com/assets/42368/26384696/d5bb7234-3fef-11e7-8602-867d270a5fc0.png)

It can auto re-connect in case of connection loss;

![](http://g.recordit.co/9xWYmLlRz2.gif)

**It doesn't require a change on existing implementation and it works well for the existing events; with a full scenario screencast can be watched from [here](http://recordit.co/7mzvJlywm3).**

## How Has This Been Tested?

It's tested manually for now, there will be integration tests for the worker and client side implementation in the near feature.

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist:
- [x] All new and existing tests passed.

gokmen committed:
NotificationWorker: initial commit

gokmen committed:
Workers: run 3 instances of notification worker on default env.

gokmen committed:
Workers: support customization on workers from options

gokmen committed:
Default env: disable dispatcher, pubnub; enable notification, emailer

gokmen committed:
Nginx: handle false state correctly for disableLocation

gokmen committed:
NotificationWorker: provide same endpoint with socialapi dispatcher

This one is a drop-in replacement for api/social/dispatcher endpoint
which will provide the same output for the given input. So, if we env
doesn't provide pubnub all the existing code can use this endpoint
instead, no further change is needed in the codebase.

gokmen committed:
Client: some debug and update point is provided for new notifications

gokmen committed:
Default: enable nginx locations for notification worker

gokmen committed:
RealTimeController: simplify initPubNub, group PubNub related calls

gokmen committed:
Environment: simplify enabledWorkers flow by definition

gokmen committed:
CredentialStore: covert node.toString if it’s binary

gokmen committed:
NodeNotificationClient: initial commit

gokmen committed:
Nginx.conf: fix cyclomatic complexity

gokmen committed:
NodeNotificationClient: emit all other messages if it’s not for auth

gokmen committed:
NotificationController: unify realtime update handlers

gokmen committed:
RealtimeController: provide NodeNotificationClient if needed

Currently ::subscribeNotification was returning a PubNubChannel which
is a simple wrapper that emits `message` over it, with this change RTC
will return NodeNotificationClient if PubNub is not supported by the
current environment. And since it’s providing the same flow with PNC, it
will work as is.

gokmen committed:
GroupsController: listen realtime updates from NodeNotificationClient

This will keep the existing channel subscription for socialApi over
PubNub as is but it will use the NNC if PN is not enabled in current env

gokmen committed:
Workers: run only 1 NotificationWorker instance by default

If it's required to have more than 1 instance of notification worker,
sticky sessions should be enabled on the load balancer if it's willing
to use long polling (xhr-polling/stream)

If client supports WS everything will work as expected with multiple NW
instances. But if client needs to fallback to long polling somehow this
will break the NW auth mechanism; current nginx load balancer is working
with round-robin and sends each request to next worker which breaks xhr.
To be able to fix that load balancer should route the all requests from
same source to same NW instance. This can be handled by using sticky
sessions. If client support WS everything will be fine since the auth is
getting handled in already open socket connection which won’t be changed
before closed.

gokmen committed:
NodeNotificationClient: try to re-connect with exponential backoff

gokmen committed:
NodeNotificationWorker: provide recipient information with messages

gokmen committed:
NodeNotificationClient: route messages based on the provided recipients

Currently groupsController is listening for group related events from
one channel and notificationController is listening one from another for
user specific events. To be able to keep it as is NNW is providing
__to information which includes user and group information as recipient.
If only group is provided which means this message needs to be handled
by groupsController.

szkl commented on client/app/lib/realtimecontroller.coffee:
```diff
@@ -61,58 +66,52 @@ module.exports = class RealtimeController extends KDController
 
-  syncTime: ->
+      @pubnub.time (timestamp) =>
+        return callback Date.now() * 10000  unless timestamp
```
Would it have any side-effect to set browser time as `@lastSeenOnline` property?

gokmen replied:
have no idea, just moved it out to same block with init flow. any idea @cihangir ?

cihangir replied:
users might have invalid times, that creates issue while fetching the message history

szkl commented on workers/notification/lib/notification/notificationworker.coffee:
```diff
@@ -0,0 +1,335 @@
+      [ ... , connectionId ] = route
+      connection.id isnt connectionId
+
+    delete @_connections[connection.id]
```
@cihangir L310?

gokmen commented on client/app/lib/nodenotificationclient.coffee:
```diff
@@ -0,0 +1,136 @@
+
+    debug 'connection established, sending auth message...'
+    @_send @_getAuth()
+    kd.utils.wait 5000, =>
```
if it ends up with connecting to sockjs server and sending the message sometimes I encountered no response issue that just happens. to prevent possible race condition there I'm waiting for 5 sec. and checking for the connection status again. otherwise backoff is handling retry part.

szkl commented on client/app/lib/nodenotificationclient.coffee:
```diff
@@ -0,0 +1,136 @@
+
+    @_socket.onmessage = @bound '_handleNewMessage'
+    @_socket.onclose   = @bound '_handleClose'
+    @_socket.onopen    = @bound '_handleOpen'
```
I think this should subscribe to `error` event too.

szkl replied:
An alternative way to bind event handlers would be calling [`EventTarget::addEventListener`](https://github.com/sockjs/sockjs-client/blob/master/lib/event/eventtarget.js#L11-L22) method. That way can easily address issues like [this](https://github.com/koding/koding/pull/11234/files#r118319648).

gokmen replied:
errors are not handled if they don't end with close so I ended up with using close only, what should we do when error happened but connection is still open?

Didn't your last comment though :) how it will help to address the issue?

szkl replied:
We can simply log the errors first and deal with handling them as we see any.

If connection ready state is not open yet then just bind an event handler and send the message in the inline callback function. Otherwise, messages need to be stored in a queue.

gokmen replied:
ah ok I got it now, you're right.

gokmen commented on client/app/lib/nodenotificationclient.coffee:
```diff
@@ -0,0 +1,136 @@
+    @readyState = 0
+    @once 'ready', => @readyState = 1
+
+    @_bo.backoff()  if event?.code > 1000
```
I'll add as a comment, but it's generic knowledge I thought :) it's simply this https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent `0-999` is not used, `1000` is normal close if some how if we implement stop notifications we can call `close` manually or if connection closed by remote (if your session is invalid etc.) we don't need/want to retry to connect. But in any other cases `> 1000` we will try to re-connect.

gokmen commented on client/app/lib/nodenotificationclient.coffee:
```diff
@@ -0,0 +1,136 @@
+
+    return  unless message = @_parse message
+
+    if message.auth is 'ok' and not message.__to?
```
I thought the same but I didn't want to conflict with existing notification schema that is provided by social api dispatcher. currently there are different channels for handling `__to` part, and I tried to find a name for it in the message as much as possible to prevent conflict. but I can make something `message.type = 'notification:auth', message.status = ok` for auth handling as suggested.

gokmen commented on client/app/lib/nodenotificationclient.coffee:
```diff
@@ -0,0 +1,136 @@
+      message = JSON.parse message.data
+    catch e
+      debug 'got a corrupted message', message
+      @close internal = yes
```
I did the same on the worker side too, but I'm not sure too. I just tried to prevent sending non standard messages (ie: JSON) between server-client. I can only log that.

gokmen commented on client/app/lib/nodenotificationclient.coffee:
```diff
@@ -0,0 +1,136 @@
+
+    return {
+      auth : {
+        clientId  : kookies.get('clientId')
```
yup, username and groupName was only for verification but you're right we don't need that query to go through db.

gokmen commented on workers/notification/lib/notification/notificationworker.coffee:
```diff
@@ -0,0 +1,335 @@
+
+    @log 'starting worker...'
+
+    @_connectToRabbitMQ()
```
we're consuming but we're not sending them through before making the bongo connection https://github.com/koding/koding/pull/11234/files#diff-8367197550fb4e5f4852ba02179f6710R70

gokmen commented on package.json:
```diff
@@ -96,6 +96,7 @@
     "sendmail": "1.1.1",
     "shortid": "2.2.6",
     "simple-recaptcha": "0.0.3",
+    "sockjs": "^0.3.18",
```
yeah why it didnt 🤔 I'll update in any case.

szkl replied:
https://github.com/koding/koding/pull/11239

szkl commented on client/app/lib/nodenotificationclient.coffee:
```diff
@@ -0,0 +1,136 @@
+
+  _send: (obj) ->
+
+    if @_socket
```
This can be improved by also checking the ready state of SockJS client.

```coffeescript
if @_socket and @_socket.readyState is SockJS.OPEN
```

gokmen committed:
Package.json: lock in the given version for sockjs

gokmen committed:
NodeNotificationWorker: wait bongo connect for mq subscription

gokmen committed:
NodeNotificationWorker: authentication only with clientId

gokmen committed:
NodeNotificationClient: use SockJS::addEventListener instead

gokmen committed:
NodeNotificationClient: add comments about mysterious pieces

gokmen committed:
NodeNotification: pass message in content and type with recipient

So, new notification message will be;

```
 type      : string [auth, message(default)]
 state     : string ('ok') # only for auth type
 recipient : object {group, [user]}
 content   : object        # whole message
```

gokmen committed:
NodeNotificationWorker: Split MQController into it's own file

gokmen committed:
NodeNotificationWorker: Split ExpressController into it's own file

gokmen commented:
all the issues resolved cc/ @cihangir @szkl
